iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/pipeconsensus/PipeConsensusSyncConnector.java

```diff
@@ -0,0 +1,529 @@
+          String.format(
+              "Failed to transfer tablet insertion event %s, because %s.",
+              tabletInsertionEvent, e.getMessage()),
+          e);
+    }
+  }
+
+  @Override
+  public void transfer(TsFileInsertionEvent tsFileInsertionEvent) throws Exception {
+    // Note: here we don't need to do type judgment here, because PipeConsensus uses DO_NOTHING
+    // processor and will not change the event type like
+    // org.apache.iotdb.db.pipe.connector.protocol.thrift.sync.IoTDBDataRegionSyncConnector
+    try {
+      // In order to commit in order
+      if (isTabletBatchModeEnabled && !tabletBatchBuilder.isEmpty()) {
+        doTransfer();
+      }
+
+      doTransfer((PipeTsFileInsertionEvent) tsFileInsertionEvent);
+    } catch (Exception e) {
+      throw new PipeConnectionException(
+          String.format(
+              "Failed to transfer tsfile insertion event %s, because %s.",
+              tsFileInsertionEvent, e.getMessage()),
+          e);
+    }
+  }
+
+  @Override
+  public void transfer(Event event) throws Exception {
+    // in order to commit in order
+    if (isTabletBatchModeEnabled && !tabletBatchBuilder.isEmpty()) {
+      doTransfer();
+    }
+
+    if (!(event instanceof PipeHeartbeatEvent)) {
+      LOGGER.warn(
+          "PipeConsensusSyncConnector does not support transferring generic event: {}.", event);
+    }
+  }
+
+  private void doTransfer() {
+    try (final SyncPipeConsensusServiceClient syncPipeConsensusServiceClient =
+        syncRetryAndHandshakeClientManager.borrowClient(getFollowerUrl())) {
+      final TPipeConsensusBatchTransferResp resp;
+      resp =
+          syncPipeConsensusServiceClient.pipeConsensusBatchTransfer(
+              tabletBatchBuilder.toTPipeConsensusBatchTransferReq());
+
+      final List<TSStatus> statusList =
+          resp.getBatchResps().stream()
+              .map(TPipeConsensusTransferResp::getStatus)
+              .collect(Collectors.toList());
+
+      // TODO(support batch): handle retry logic
+      // Only handle the failed statuses to avoid string format performance overhead
+      //      if (status.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()
+      //          && status.getCode() != TSStatusCode.REDIRECTION_RECOMMEND.getStatusCode()) {
+      //        receiverStatusHandler.handle(
+      //            resp.getStatus(),
+      //            String.format(
+      //                "Transfer PipeConsensusTransferTabletBatchReq error, result status %s",
+      //                resp.status),
+      //            tabletBatchBuilder.deepCopyEvents().toString());
+      //      }
+
+      tabletBatchBuilder.onSuccess();
+    } catch (Exception e) {
+      throw new PipeConnectionException(
+          String.format(
+              PIPE_CONSENSUS_SYNC_CONNECTION_FAILED_FORMAT,
+              getFollowerUrl().getIp(),
+              getFollowerUrl().getPort(),
+              e.getMessage(),
+              TABLET_BATCH_SCENARIO),
```

Comment:
Check the sequence here... PipeConsensus: syncClient connection to %s:%s failed when exception, because: transfer tablet batch.. And is same for other exception messages.

Reply:
good catch, fixed